submariner/OVN-Interconnect.md

```diff
@@ -0,0 +1,239 @@
+  Submariner uses for the logical router. In this proposal, we plan to remove the submariner logical router and the associated switches and
+  this will not be a problem anymore.
+
+  The below indicates node annotations in a single zone OVN setup and is added by OVN.
+
+```bash
+    annotations:
+      k8s.ovn.org/ovn-node-transit-switch-port-ips: '["169.254.0.1/16"]'
+      k8s.ovn.org/ovn-zone: global
+    name: cluster1-worker
+
+    annotations:
+      k8s.ovn.org/ovn-node-transit-switch-port-ips: '["169.254.0.1/16"]'
+      k8s.ovn.org/ovn-zone: global
+    name: cluster2-worker
+```
+
+* Multiple Zone: In a multiple-zone setup, we will have an OVN database and an OVN master pod for each zone. Transit switches connect the
+  zones. The OVN-Kubernetes services ensure that the necessary routes are added for pod and service reachability across nodes in different
+  zones.
+
+```bash
+    annotations:
+      k8s.ovn.org/ovn-node-transit-switch-port-ips: '["169.254.0.3/16"]'
+      k8s.ovn.org/ovn-zone: global
+    name: cluster1-worker
+
+    annotations:
+      k8s.ovn.org/ovn-node-transit-switch-port-ips: '["169.254.0.5/16"]'
+      k8s.ovn.org/ovn-zone: az2
+    name: cluster2-worker
+```
+
+With the current architecture, Submariner adds routes only in the zone in which it is deployed. For example, if Submariner is deployed in
+zone 1 it programs OVN db in zone 1. So only pods in zone 1 nodes will be able to talk to other clusters. Pods in zone 2 or zone 3 will not
+be able to reach remote clusters connected via Submariner. With these changes we plan to support both the modes and cluster where interconnect
+is not enabled as well.
+
+### Route APIs
+
+As part of this proposal, we are planning to add two new CRDs in Submariner. The plan is to decouple OVN dataplane programming so that this
+can be moved to OVN, when OVN provides a way to add custom routes. We can propose this CRD or consume the CRD that they provide at a later
+point of time.
+
+#### SubmarinerGwRoute
+
+This CR will be created when a remote endpoint is added and there will be one CR per remote cluster. This crd has two fields
+
+* NextHop - Specifies the next hop to reach the remote cluster, in this case it will be the IP of ovn-k8s-mp0
+  interface, the interface used by OVN for host networking.
+* RemoteCIDR - Specifies the list of  remote CIDRs reachable via this cluster.
+
+This CR will be used by the route agent pod running on the non-Gateway nodes to program OVN to send the traffic destined to
+remote clusters via the Submariner tunnel.
+
+``` go
+type SubmarinerGwRoute struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    SubmarinerRoutePolicySpec SubmarinerRoutePolicySpec `json:"submarinerRoutePolicySpec"`
+}
+
+type SubmarinerRoutePolicySpec struct {
+    //Specifies the next hop to reach the remote CIDRs
+    NextHop string `json:"nextHop"`
```

Comment:
Shouldn't 'NextHop' be a list? to enable multiple routes to the destination (ECMP) ?

Reply:
may be  we can change it . @sridhargaddam Do you agree?

Reply:
If I understand correctly, this is for the `SubmarinerGwRoute` CR. The nextHop here points to the IPaddress on `ovn-k8s-mp0` of the GW node. Assuming that we support Active/Active HA in future, can we use the same CR and configure multiple NextHops which point to `ovn-k8s-mp0` IPs of different GW nodes (and should they be all in the same Zone or can we even use the IPs if they are in different zone)?

Reply:
yes that is right , if we want a next hop for a set of remote CIDRS , it can be achieved with this CR.

Reply:
Not sure that I fully understood,  you suggest that for Active/Active  we should create 2 SubmarinerGwRoute CRs?

Reply:
made it an array.